middle_end/flambda/compilenv_deps/target_imm.ml

```diff
@@ -5,243 +5,289 @@
+(*   Copyright 2014--2021 Jane Street Group LLC                           *)
 (*                                                                        *)
 (*   All rights reserved.  This file is distributed under the terms of    *)
 (*   the GNU Lesser General Public License version 2.1, with the          *)
 (*   special exception on linking described in the file LICENSE.          *)
 (*                                                                        *)
 (**************************************************************************)
 
-[@@@ocaml.warning "+a-4-30-40-41-42"]
+[@@@ocaml.warning "+a-30-40-41-42"]
 
-module TO = Targetint.OCaml
+let () =
+  match Targetint.num_bits with
+  | Sixty_four -> ()
+  | Thirty_two ->
+    if Config.flambda then begin
+      Misc.fatal_error "Flambda 2 does not yet support 32-bit compilation"
+    end
+
+module Imm = struct
+  module T0 = struct
+    include Numbers.Int64
+
+    let compare = compare
+
+    let equal = equal
+
+    let hash = hash
+
+    let output = output
+
+    let print = print
+
+    let minus_one = -1L
+
+    let zero = 0L
+
+    let one = 1L
+
+    let ten = 10L
+
+    let hex_ff = 0xffL
+
+    let min_value = Int64.min_int
+
+    let max_value = Int64.max_int
+
+    let sub = Int64.sub
+
+    let neg = Int64.neg
+
+    let shift_left = Int64.shift_left
+
+    let shift_right = Int64.shift_right
+
+    let shift_right_logical = Int64.shift_right_logical
+
+    let xor = Int64.logxor
+
+    let or_ = Int64.logor
+
+    let and_ = Int64.logand
+
+    let mod_ = Int64.rem
+
+    let div = Int64.div
+
+    let mul = Int64.mul
+
+    let add = Int64.add
+
+    let bottom_byte_to_int t = Int64.to_int (Int64.logand t hex_ff)
+
+    let of_char c = Int64.of_int (Char.code c)
+
+    let of_int = Int64.of_int
+
+    let to_int = Int64.to_int
+
+    let of_int_option i = Some (of_int i)
+
+    let of_int32 t = Int64.of_int32 t
+
+    let to_int32 t = Int64.to_int32 t
+
+    let of_int64 t = t
+
+    let to_int64 t = t
+
+    let of_float = Int64.of_float
+
+    let to_float = Int64.to_float
+
+    let to_targetint = Targetint.of_int64
+
+    let of_targetint = Targetint.to_int64
+
+    let max_array_length = Int64.sub (Int64.shift_left 1L 54) 1L
+
+    let max_string_length = Int64.sub (Int64.mul 8L max_array_length) 1L
+
+    let max t1 t2 = if Stdlib.compare t1 t2 < 0 then t2 else t1
+
+    let min t1 t2 = if Stdlib.compare t1 t2 < 0 then t1 else t2
+
+    let ( <= ) t1 t2 = Stdlib.( <= ) (compare t1 t2) 0
+
+    let ( < ) t1 t2 = Stdlib.( < ) (compare t1 t2) 0
+
+    let to_int_option t =
+      let min_int_as_int64 = Int64.of_int Stdlib.min_int in
+      let max_int_as_int64 = Int64.of_int Stdlib.max_int in
+      if min_int_as_int64 <= t && t <= max_int_as_int64 then Some (to_int t)
+      else None
+
+    let to_int_exn t =
+      match to_int_option t with
+      | Some i -> i
+      | None -> Misc.fatal_errorf "Target_imm.Imm.to_int_exn %Ld" t
```

Comment:
Maybe add an "out of range" somewhere in the error message ?

Reply:
Fixed